website/project/views/comment.py

```diff
@@ -13,106 +14,169 @@
+@must_be_contributor_or_public
+def view_comments_project(auth, **kwargs):
+    """
+    Returns information needed to get comments for the total discussion page
+    """
+
+    node = kwargs['node'] or kwargs['project']
+    page = request.args.get('page', None)
+    root_id = node._id if page == Comment.OVERVIEW else None
+
+    if page:
+        _update_comments_timestamp(auth, node, page=page, root_id=root_id)
+
+    ret = {
+        'comment_target': page or 'total',
+        'comment_target_id': root_id
+    }
+    ret.update(_view_project(node, auth, primary=True, check_files=(not page)))
+    return ret
+
+
+@must_be_contributor_or_public
+def view_comments_single(auth, **kwargs):
+    """
+    Returns information needed to get a single comment and its replies
+    """
+    node = kwargs['node'] or kwargs['project']
+    comment = kwargs_to_comment(kwargs)
+    serialized_comment = serialize_comment(comment, auth)
 
+    from website.addons.wiki.model import NodeWikiPage
+
+    ret = {
+        'comment': serialized_comment,
+        'comment_target': serialized_comment['page'],
+        'comment_target_id': comment.root_target.page_name
+        if isinstance(comment.root_target, NodeWikiPage)
+        else comment.root_target._id
+    }
+    ret.update(_view_project(node, auth, primary=True))
+    return ret
+
+
+def resolve_target(node, page, guid):
     if not guid:
         return node
     target = Guid.load(guid)
     if target is None:
+        if page == Comment.WIKI:
+            return node.get_wiki_page(guid, 1)
         raise HTTPError(http.BAD_REQUEST)
     return target.referent
 
 
-def collect_discussion(target, users=None):
-
-    users = users or collections.defaultdict(list)
-    for comment in getattr(target, 'commented', []):
-        if not comment.is_deleted:
-            users[comment.user].append(comment)
-        collect_discussion(comment, users=users)
+def collect_discussion(target, users):
+    if not getattr(target, 'commented', None) is None:
+        for comment in getattr(target, 'commented', []):
+            if not (comment.is_deleted or comment.is_hidden):
+                users[comment.user].append(comment)
+            collect_discussion(comment, users=users)
     return users
 
 
-@must_be_contributor_or_public
-def comment_discussion(**kwargs):
+def comment_discussion(comments, node, anonymous=False, widget=False):
 
-    node = kwargs['node'] or kwargs['project']
-    auth = kwargs['auth']
-    users = collect_discussion(node)
-    anonymous = has_anonymous_link(node, auth)
-    # Sort users by comment frequency
-    # TODO: Allow sorting by recency, combination of frequency and recency
-    sorted_users = sorted(
+    users = collections.defaultdict(list)
+    for comment in comments:
+        if not (comment.is_deleted or comment.is_hidden):
+            users[comment.user].append(comment)
+        if not widget:
+            collect_discussion(comment, users=users)
```

Comment:
`collect_discussion` modifies the `users` dict in-place, which makes this logic hard to follow. In general, we try to avoid in-place modifications when possible.

How about refactoring `collect_discussion` to build and return the dictionary, so the signature becomes:

```
collect_discussion(<list of comments>) -> <dict of user: <list of comments>  pairs>
```